### PR TITLE
View comma separated values for SelectMultiple in Hierarchy  view

### LINKED ISSUE #1839: Selected options from select multiple question are not separated in hierarchy view.
#### Software and hardware versions 
Collect v1.12.2 and older version

#### Problem description
Selected options from select multiple question are not separated in hierarchy view.

#### Steps to reproduce the problem
1. Use any form with select multiple question.
2. Select at least two options.
3. Go to hierarchy view.

#### Expected behavior
Selected options from select multiple question should be separated by a comma or a semicolon in hierarchy view.

### PR DESCRIPTION
**Pull request by akshay-ap**
Closes #1839

#### What has been done to verify that this works as intended?
Tested on Moto M (Api 24)

#### Why is this the best possible solution? Were any other approaches considered?
Other approach needed the change in Javarosa library. But, doing so might affect other apps.
So, a custom string is built on Collect to show `,` separated values.

#### Are there any risks to merging this code? If so, what are they?
No.

#### Do we need any specific form for testing your changes? If so, please attach one.
Yes. Needs testing on different forms.

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/utilities/FormEntryPromptUtils.java [hunk 1]
```diff
@@ -37,6 +40,18 @@ public static String getAnswerText(FormEntryPrompt fep, Context context) {
...
+        if (data instanceof SelectMultiData) {
+            String text;
+            StringBuilder b = new StringBuilder();
+            List<Selection> values = (List<Selection>) data.getValue();
+            for (Selection value : values) {
+                b.append(fep.getSelectItemText(value)).append(", ");
+            }
+            b.delete(b.length() - 2, b.length());
+            text = b.toString();
+            return text;
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/utilities/FormEntryPromptUtils.java]: you can just return `b.toString()` then you don't need `text` at all.

File: collect_app/src/main/java/org/odk/collect/android/utilities/FormEntryPromptUtils.java [hunk 2]
```diff
@@ -37,6 +40,18 @@ public static String getAnswerText(FormEntryPrompt fep, Context context) {
         IAnswerData data = fep.getAnswerValue();
         final String appearance = fep.getQuestion().getAppearanceAttr();
 
+        if (data instanceof SelectMultiData) {
+            String text;
+            StringBuilder b = new StringBuilder();
+            List<Selection> values = (List<Selection>) data.getValue();
+            for (Selection value : values) {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/utilities/FormEntryPromptUtils.java]: Here I would just use `for (Selection value : (List<Selection>) data.getValue()) {` insteadof creating `values` list which is used only here.

File: collect_app/src/main/java/org/odk/collect/android/utilities/FormEntryPromptUtils.java [hunk 3]
```diff
@@ -37,6 +40,18 @@ public static String getAnswerText(FormEntryPrompt fep, Context context) {
...
+                b.append(fep.getSelectItemText(value)).append(", ");
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/utilities/FormEntryPromptUtils.java]:
```
b
   .append(fep.getSelectItemText(value))
   .append(", ");
```

would be more legible.

File: collect_app/src/main/java/org/odk/collect/android/utilities/FormEntryPromptUtils.java (hunk 2 above)
Comment (akshay-ap) [reply collect_app/src/main/java/org/odk/collect/android/utilities/FormEntryPromptUtils.java]: Done

File: collect_app/src/main/java/org/odk/collect/android/utilities/FormEntryPromptUtils.java (hunk 3 above)
Comment (akshay-ap) [reply collect_app/src/main/java/org/odk/collect/android/utilities/FormEntryPromptUtils.java]: Done

File: collect_app/src/main/java/org/odk/collect/android/utilities/FormEntryPromptUtils.java (hunk 1 above)
Comment (akshay-ap) [reply collect_app/src/main/java/org/odk/collect/android/utilities/FormEntryPromptUtils.java]: Done

**Comment by grzesiek2010** [comment]
Looks good to me thanks. 

We need to know what @yanokwa and @lognaturel think. Maybe you prefer another separator like semicolon? If it's ok for you it's ready for testing.

**Review by yanokwa — DISMISSED** [review]
Really loving the back and forth here! Perhaps we can make this even simpler?

```java
String sep = "";
for (Selection value : (List<Selection>) data.getValue()) {
  b.append(sep);
  sep = ", ";
  b.append(fep.getSelectItemText(value));
}
return b.toString();
```

**Review by shobhitagarwal1612 — APPROVED** [review]

**Comment by lognaturel** [comment]
Thanks for taking this on, @akshay-ap! Since you're now a veteran contributor and community member, please leave `quick win` issues for people who are less experienced! 😊

**Comment by akshay-ap** [comment]
Sure !!! 😀😀

**Review by grzesiek2010 — APPROVED** [review]

**Comment by mmarciniak90** [comment]
**Tested with success!**
Verified on Android: 4.1, 4..2, 4.4, 5.1, 6.0, 7.0, 8.0

|  |  |

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

**Review by lognaturel — APPROVED** [review]

**Pull request merged**